File: app/static/js/password_policy.js
/**
 * Client-side password policy (mirrors app/security/password_policy.py).
 */
(function (global) {
  const MIN_LENGTH = 8;
  const MAX_LENGTH = 128;
  const SPECIAL = /[!@#$%^&*()_+\-=[\]{}|;:'",.<>?/\\`~]/;

  const COMMON = new Set([
    "password", "password1", "password12", "password123",
    "12345678", "123456789", "1234567890", "qwerty", "qwerty123",
    "admin", "admin123", "letmein", "welcome", "welcome1",
    "iloveyou", "monkey", "dragon", "master", "abc123", "11111111",
    "aura", "aura123", "customer", "test1234", "changeme",
  ]);

  const MIN_SEQUENCE_RUN = 6;

  const SEQUENCES = [
    "0123456789", "9876543210",
    "abcdefghijklmnopqrstuvwxyz", "zyxwvutsrqponmlkjihgfedcba",
    "qwertyuiop", "asdfghjkl", "zxcvbnm",
  ];

  function hasSequence(lower) {
    const n = MIN_SEQUENCE_RUN;
    for (const seq of SEQUENCES) {
      for (let i = 0; i <= seq.length - n; i++) {
        if (lower.includes(seq.slice(i, i + n))) return true;
      }
    }
    return false;
  }

  function checkRule(password, username, email) {
    return {
      length: password.length >= MIN_LENGTH && password.length <= MAX_LENGTH,
      upper: /[A-Z]/.test(password),
      lower: /[a-z]/.test(password),
      digit: /\d/.test(password),
      special: SPECIAL.test(password),
      noSpace: !/\s/.test(password),
      notCommon: !COMMON.has(password.toLowerCase()),
      notRepeat: !/(.)\1{3,}/.test(password),
      notSeq: !hasSequence(password.toLowerCase()),
      notPersonal: (() => {
        const lower = password.toLowerCase();
        const u = (username || "").trim().toLowerCase();
        if (u.length >= 3 && lower.includes(u)) return false;
        const local = ((email || "").split("@")[0] || "").trim().toLowerCase();
        if (local.length >= 3 && lower.includes(local)) return false;
        return true;
      })(),
    };
  }

  function validatePassword(password, username, email) {
    if (!password) {
      return { ok: false, message: "Password is required." };
    }
    const r = checkRule(password, username, email);
    if (!r.length) {
      return { ok: false, message: `Password must be ${MIN_LENGTH}–${MAX_LENGTH} characters.` };
    }
    if (!r.noSpace) return { ok: false, message: "Password cannot contain spaces." };
    if (!r.upper) return { ok: false, message: "Add at least one uppercase letter (A–Z)." };
    if (!r.lower) return { ok: false, message: "Add at least one lowercase letter (a–z)." };
    if (!r.digit) return { ok: false, message: "Add at least one number." };
    if (!r.special) {
      return { ok: false, message: "Add at least one special character (e.g. ! @ # $)." };
    }
    if (!r.notCommon) {
      return { ok: false, message: "This password is too common. Choose a stronger one." };
    }
    if (!r.notRepeat) {
      return { ok: false, message: "Avoid four or more identical characters in a row." };
    }
    if (!r.notSeq) {
      return {
        ok: false,
        message: `Avoid ${MIN_SEQUENCE_RUN}+ characters in a row from a simple sequence (e.g. 123456 or qwerty).`,
      };
    }
    if (!r.notPersonal) {
      return { ok: false, message: "Password cannot include your username or email." };
    }
    return { ok: true, message: "", rules: r };
  }

  function bindPasswordChecker(opts) {
    const pwd = document.querySelector(opts.passwordSelector);
    const confirm = opts.confirmSelector
      ? document.querySelector(opts.confirmSelector)
      : null;
    const list = opts.checklistSelector
      ? document.querySelector(opts.checklistSelector)
      : null;
    const usernameEl = opts.usernameSelector
      ? document.querySelector(opts.usernameSelector)
      : null;
    const emailEl = opts.emailSelector
      ? document.querySelector(opts.emailSelector)
      : null;

    if (!pwd || !list) return;

    const items = {
      length: list.querySelector('[data-rule="length"]'),
      upper: list.querySelector('[data-rule="upper"]'),
      lower: list.querySelector('[data-rule="lower"]'),
      digit: list.querySelector('[data-rule="digit"]'),
      special: list.querySelector('[data-rule="special"]'),
      match: list.querySelector('[data-rule="match"]'),
    };

    function paint() {
      const username = usernameEl ? usernameEl.value.trim() : "";
      const email = emailEl ? emailEl.value.trim() : "";
      const r = checkRule(pwd.value, username, email);
      const set = (el, ok) => {
        if (!el) return;
        el.classList.toggle("pwd-ok", ok);
        el.classList.toggle("pwd-bad", !ok);
      };
      set(items.length, r.length && r.noSpace);
      set(items.upper, r.upper);
      set(items.lower, r.lower);
      set(items.digit, r.digit);
      set(items.special, r.special);
      if (items.match && confirm) {
        const match =
          confirm.value.length > 0 && pwd.value === confirm.value;
        set(items.match, match);
      }
    }

    ["input", "change"].forEach((ev) => {
      pwd.addEventListener(ev, paint);
      if (confirm) confirm.addEventListener(ev, paint);
      if (usernameEl) usernameEl.addEventListener(ev, paint);
      if (emailEl) emailEl.addEventListener(ev, paint);
    });
    paint();
  }

  global.AuraPasswordPolicy = {
    MIN_LENGTH,
    validatePassword,
    bindPasswordChecker,
  };
})(window);

File: app/security/password_policy.py
"""Password strength rules for registration and password changes."""

from __future__ import annotations

import re

MIN_LENGTH = 8
MAX_LENGTH = 128

_SPECIAL_RE = re.compile(r'[!@#$%^&*()_+\-=\[\]{}|;:\'",.<>?/\\`~]')

_COMMON_PASSWORDS = frozenset({
    "password",
    "password1",
    "password12",
    "password123",
    "12345678",
    "123456789",
    "1234567890",
    "qwerty",
    "qwerty123",
    "qwertyuiop",
    "admin",
    "admin123",
    "letmein",
    "welcome",
    "welcome1",
    "iloveyou",
    "monkey",
    "dragon",
    "master",
    "sunshine",
    "princess",
    "football",
    "shadow",
    "abc123",
    "11111111",
    "00000000",
    "aura",
    "aura123",
    "customer",
    "test1234",
    "changeme",
})

_SEQUENCES = (
    "0123456789",
    "9876543210",
    "abcdefghijklmnopqrstuvwxyz",
    "zyxwvutsrqponmlkjihgfedcba",
    "qwertyuiop",
    "asdfghjkl",
    "zxcvbnm",
)


def password_requirements_hint() -> str:
    return (
        "Use at least 8 characters including uppercase, lowercase, a number, "
        "and a special character (!@#$…). Avoid common passwords and your username/email."
    )


def validate_password_strength(
    password: str,
    *,
    username: str = "",
    email: str = "",
) -> str | None:
    """Return an error message, or None if the password meets policy."""
    if not password:
        return "Password is required."

    if len(password) < MIN_LENGTH:
        return f"Password must be at least {MIN_LENGTH} characters."
    if len(password) > MAX_LENGTH:
        return f"Password must be at most {MAX_LENGTH} characters."
    if re.search(r"\s", password):
        return "Password cannot contain spaces."
    if not re.search(r"[A-Z]", password):
        return "Password must include at least one uppercase letter (A–Z)."
    if not re.search(r"[a-z]", password):
        return "Password must include at least one lowercase letter (a–z)."
    if not re.search(r"\d", password):
        return "Password must include at least one number."
    if not _SPECIAL_RE.search(password):
        return "Password must include at least one special character (e.g. ! @ # $ %)."

    lowered = password.lower()
    if lowered in _COMMON_PASSWORDS:
        return "This password is too common. Choose a stronger, unique password."

    if re.search(r"(.)\1{3,}", password):
        return "Password cannot contain four or more identical characters in a row."

    if _contains_simple_sequence(lowered):
        return "Password cannot be a simple keyboard or numeric sequence."

    uname = (username or "").strip().lower()
    if len(uname) >= 3 and uname in lowered:
        return "Password cannot contain your username."

    if email:
        local = email.split("@", 1)[0].strip().lower()
        if len(local) >= 3 and local in lowered:
            return "Password cannot contain part of your email address."

    return None


def _contains_simple_sequence(lower_password: str) -> bool:
    for seq in _SEQUENCES:
        for i in range(len(seq) - 3):
            if seq[i : i + 4] in lower_password:
                return True
    return False
